### ISSUE TITLE
Dynamic GPU support

### ISSUE THREAD

**Issue by LukeShortCloud** [LukeShortCloud/winesapOS#721]
Handle switching between different GPU drivers (ex., proprietary NVIDIA to open source AMD) in an automated way.

Since the LightDM fail-safe measure has been added by @wallentx , I am opening up a separate GitHub Issue for tracking dynamic GPU support.

Original discussion: https://github.com/LukeShortCloud/winesapOS/issues/670

**Comment by LukeShortCloud** [comment]
@wallentx What are you thoughts on relying on the nouveau (OpenGL) and NVK (Vulkan) drivers instead of the proprietary ones? I know it'll take some time for NVK to mature but I have heard that NVK is about to support Vulkan 1.3 (it is only 1 Vulkan extension away!) which will fully support Proton. It is still slow but the developers said performance will be the next thing they tackle. Eventually Zink (OpenGL-on-Vulkan) will provide an even better OpenGL experience, too.

In summary, it sounds like this issue will be solved organically upstream this year with NVIDIA getting great out-of-the-box Linux and Mesa support. Do you think that having dynamic GPU support for the proprietary NVIDIA drivers is still worth looking into? Curious to hear your thoughts. I'm open to either way.

Related: https://github.com/LukeShortCloud/winesapOS/issues/750

**Comment by wallentx** [comment]
Hey! I've been out of the country for the past few weeks, and my computer life has been on hold since then. I haven't personally tried NVK, but saw something either on Phoronix, or on Brodie Robertson's YouTube channel about it looking potentially promising, but current state wasn't anywhere competitive just yet. I'd have to look into it and see what it's like.

As for the dynamic GPU support, I assume you are talking about shipping the OS with all the drivers needed, rather than prompting the user to install AMD (no-op), or Nvidia drivers, correct? With the way that I've been using it, I initially set it up with Nvidia drivers, but once those were installed, I've been able to toggle between cards with no problem, so long as I don't have an xorg.conf file saved.

Another thing that might be a solution, is the `nvidia-open-dkms` drivers.. the `nvidia-open` drivers do not work with `nvidia-settings`, which is where any sort of unique xorg.conf file would get generated from, so that would at least eliminate that possibility.

**Comment by LukeShortCloud** [comment]
No problem about the delayed response. I hope you had a great time wherever you went!

My proposal is to not waste time on creating a script or program for dynamic GPU switching. Mesa 23.3 now ships the new NVK driver which, combined with the Linux 6.7 kernel which has NVIDIA GSP firmware support, will be a no-op for end-users. Is it a fast driver today? No, not yet and it's getting there. The NVK developers said 2023 was about getting it working and 2024 is about making it fast. winesapOS will still provide the proprietary driver and legacy drivers as an option for a full range of hardware support. Anyone with a >= 1600 series NVIDIA card can use NVK which should work seamlessly across different computers without any Xorg configuration changes.

I have extensively tested `nvidia-open` a few months ago. It was awful. 9 out of 10 times it did not work (failed to build, failed to load, weird errors, etc.). With that being said, the new NVIDIA 550 driver was just released and they have said that `nvidia-open` quality has been promoted from "alpha" to "beta". Either way, it is missing features like CUDA so you are better off either (1) using the proprietary driver which is stable and supports CUDA or (2) use NVK for a seamlessly "just works" experience.

tl;dr? I think NVIDIA support in Linux will finally not suck by the end of this year. :-)
